### PR TITLE
Windows Proxy Agent: Adding error handling to get_memory_in_mb()

### PR DESCRIPTION
**Pull request by shahneerali**

### CODE REVIEW

File: proxy_agent/src/common/windows.rs [hunk 1]
```diff
@@ -19,13 +19,16 @@ pub fn get_processor_count() -> usize {
...
-pub fn get_memory_in_mb() -> u64 {
+pub fn get_memory_in_mb() -> Result<u64, String> {
     let mut data = MaybeUninit::<MEMORYSTATUSEX>::uninit();
     let data = data.as_mut_ptr();
     unsafe {
         (*data).dwLength = std::mem::size_of::<MEMORYSTATUSEX>() as u32;
-        GlobalMemoryStatusEx(data);
-        (*data).ullTotalPhys / 1024 / 1024
+        if GlobalMemoryStatusEx(data) == 0 {
+            return Err("GlobalMemoryStatusEx failed".to_string());
```
Comment (ZhidongPeng) [inline proxy_agent/src/common/windows.rs]: https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-globalmemorystatusex
If the function fails, the return value is zero. To get extended error information, call [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror).

Please log this Error::last_os_error() too.

File: proxy_agent/src/common/helpers.rs [hunk 2]
```diff
@@ -14,7 +14,16 @@ use super::windows;
 static CURRENT_SYS_INFO: Lazy<(u64, usize)> = Lazy::new(|| {
     #[cfg(windows)]
     {
-        let ram_in_mb = windows::get_memory_in_mb();
+        let ram_in_mb = match windows::get_memory_in_mb() {
+            Ok(ram) => ram,
+            Err(e) => {
+                Error::new(
```
Comment (ZhidongPeng) [inline proxy_agent/src/common/helpers.rs]: what's this error used here?

File: proxy_agent/src/common/windows.rs [hunk 3]
```diff
@@ -52,7 +55,11 @@ mod tests {
     #[test]
     fn get_memory_in_mb_test() {
         let memory = super::get_memory_in_mb();
-        println!("Memory in MB: {}", memory);
-        assert_ne!(0, memory, "Memory cannot be 0.");
+        match memory {
+            Ok(memory) => {
+                assert_ne!(0, memory, "Memory cannot be 0.");
+            }
+            Err(e) => panic!("Failed to get memory: {}", e),
```
Comment (ZhidongPeng) [inline proxy_agent/src/common/windows.rs]: donot use panic! in test, use `assert!(false, format!("Failed to get memory: {}", e));`

File: proxy_agent/src/common/helpers.rs (hunk 2 above)
Comment (ZhidongPeng) [reply proxy_agent/src/common/helpers.rs]: we need figure out how to surface this error first.

**Review by ZhidongPeng — CHANGES_REQUESTED** [review]

File: proxy_agent/src/common/helpers.rs (hunk 2 above)
Comment (shahneerali) [reply proxy_agent/src/common/helpers.rs]: by surface do you mean logging it?

File: proxy_agent/src/common/windows.rs (hunk 1 above)
Comment (shahneerali) [reply proxy_agent/src/common/windows.rs]: adding logger class to the file and log both errors

File: proxy_agent/src/common/windows.rs [hunk 4]
```diff
@@ -19,13 +20,18 @@ pub fn get_processor_count() -> usize {
...
     let mut data = MaybeUninit::<MEMORYSTATUSEX>::uninit();
     let data = data.as_mut_ptr();
     unsafe {
         (*data).dwLength = std::mem::size_of::<MEMORYSTATUSEX>() as u32;
-        GlobalMemoryStatusEx(data);
-        (*data).ullTotalPhys / 1024 / 1024
+        if GlobalMemoryStatusEx(data) == 0 {
+            logger::write_error("GlobalMemoryStatusEx failed".to_string());
+            logger::write_error(Error::last_os_error().to_string());
+            return Err(String::from("GlobalMemoryStatusEx failed"));
```
Comment (shahneerali) [inline proxy_agent/src/common/windows.rs]: do notcall logger, add the last_os_error to exception

File: proxy_agent/src/common/helpers.rs (hunk 2 above)
Comment (shahneerali) [reply proxy_agent/src/common/helpers.rs]: write_error and look into exception for lazy func

Comment (shahneerali) [reply proxy_agent/src/common/helpers.rs]: if we wrap the function in a lazy_static!, we can use match when this function is called to raise exception.

**Review by ZhidongPeng — APPROVED** [review]

**Pull request merged**